Intro and explanation of Streams feature (#2362)
tarantool/doc: doc/book/box/stream.rst

..  _box_stream:

Streams and interactive transactions
====================================

Overview
--------

Since :tarantool-release:`2.10.0-beta1`, streams and interactive transactions
in them are implemented in iproto.

..  glossary::

    Stream
        A stream is a concept that supports multiplexing several transactions over one connection.

>         It is a more general approach that allows the implementation

Reviewer: I would add to this the information that all requests in the stream are executed strictly sequentially

In ``net.box``, a stream is an object above the connection that has the same methods
but allows executing requests sequentially.
The ID is generated on the client side automatically.
If a user writes their own connector and wants to use streams,
they must transmit the ``stream_id`` over the iproto protocol.

Interaction between streams and transactions
--------------------------------------------------------

As each stream can start a transaction, several transactions can be multiplexed over one connection.
There are multiple ways to begin, commit, and roll back a transaction.
One can do that using the appropriate stream methods, ``call``, ``eval``,
or ``execute`` with the SQL transaction syntax. Users can mix these methods.
For example, one might start a transaction using ``stream:begin()``,
and commit it with ``stream:call('box.commit')`` or ``stream:execute('COMMIT')``.
All the requests between ``stream:begin()`` and ``stream:commit()`` are executed within the same transaction.
If any request fails during the transaction, it will not affect the other requests in the transaction.
If a disconnect occurs while there is an active transaction in the stream,

> that transaction will be rolled back if it hasn't been committed before the connection failure.

Reviewer: I would add here simple code example:
local conn = net_box.connect(remote_server_addr)
local conn_space = conn.space.test  -- For example space test is exist on remote server.
local stream = conn:new_stream()
local stream_space = stream.space.test -- For example space test is exist on remote server.
stream:begin() -- Begin transaction over iproto streams
space:replace({1})
-- Empty select, transaction was not commited and
-- is not visible from requests not belonging to the
-- transaction.
conn_space:select{} 
-- Select return tuple, which was previously inserted,
-- because this select belongs to transaction.
stream_space:select({})
stream:commit() -- Commit transaction
-- Now this select also return tuple, because transaction was commited.
conn_space:select{} 
For more exaples you can look at 'test/box/net.box_iproto_transactions_over_streams.test.lua' in tarantool source code

..  _box_stream:

Streams and interactive transactions
====================================

..  _box_stream-overview:

Overview
--------

Since :tarantool-release:`2.10.0-beta1`, iproto implements streams and interactive transactions.

..  glossary::

    Stream
        A stream is a concept that supports multiplexing several transactions over one connection.

>         All requests in the stream are executed strictly sequentially.
>         It is a more general approach that allows the implementation
>         of :term:`interactive transactions <interactive transaction>`.

Reviewer 2: ```suggestion
        All requests in the stream are executed strictly sequentially.
        This approach allows the implementation
        of :term:`interactive transactions <interactive transaction>`.
```
a question comes up, "why is this approach more general?"

Unlike a thread associated with multitasking and execution within a program,
a stream occurs between a client and a server.

> It deals with connecting and transferring a protocol over the network.

Reviewer 2: now it looks like the protocol itself is transferred. but is it so? maybe it is the _data_ that is transferred via the protocol? I think the sentence needs to be rephrased

..  glossary::

    Interactive transaction
        An interactive transaction is a transaction that does not need to be sent in one request.
        ``begin``, ``commit`` and TX statements can be sent and executed in different requests.

..  _box_stream-features:

New features
------------

The primary purpose of :term:`streams <stream>` is to execute transactions via iproto.
Every stream has its own identifier, which is unique within the connection.
All requests with the same non-zero stream ID belong to the same stream.
All requests in the stream are processed synchronously.
The next request will not start executing until the previous one is completed.
If a request's stream ID is ``0``, it does not belong to any stream and is processed in the old way.

In :doc:`net.box </reference/reference_lua/net_box>`, a stream is an object above the connection that has the same methods
but allows executing requests sequentially.
The ID is generated on the client side automatically.
If a user writes their own connector and wants to use streams,
they must transmit the ``stream_id`` over the iproto protocol.

..  _box_stream-interaction:

> Interaction between streams and transactions
> --------------------------------------------------------

Reviewer 2: ```suggestion
Interaction between streams and transactions
--------------------------------------------
```

For example, one might start a transaction using ``stream:begin()``,
and commit it with ``stream:call('box.commit')`` or ``stream:execute('COMMIT')``.
All the requests between ``stream:begin()`` and ``stream:commit()`` are executed within the same transaction.
If any request fails during the transaction, it will not affect the other requests in the transaction.
If a disconnect occurs while there is an active transaction in the stream,
that transaction will be rolled back if it hasn't been committed before the connection failure.

Example:

.. code-block:: lua

    local conn = net_box.connect(remote_server_addr)
    local conn_space = conn.space.test
    local stream = conn:new_stream()
    local stream_space = stream.space.test

    -- Begin transaction over iproto streams:
    stream:begin()
    space:replace({1})

    -- Empty select, the transaction was not committed.
    -- It is not visible from the requests which do not belong to the
    -- transaction.

    -- Select return tuple, which was previously inserted,
    -- because this select belongs to transaction:
    conn_space:select{}
    stream_space:select({})

    -- Commit transaction:
    stream:commit()

>     -- Now this select also return tuple, because the transaction has been committed:

Reviewer 2: ```suggestion
    -- Now this select also returns the tuple, because the transaction has been committed:
```
if I understood it correctly

..  _box_stream:

Streams and interactive transactions
====================================

..  _box_stream-overview:

Overview
--------

> Since :tarantool-release:`2.10.0-beta1`, streams and interactive transactions
> in them are implemented in iproto.

Reviewer 2: ```suggestion
Since :tarantool-release:`2.10.0-beta1`, iproto implements streams and interactive transactions
over them.
```

..  glossary::

    Stream

>         A stream is a concept that supports multiplexing several transactions over one connection.

Reviewer 2: ```suggestion
        A stream supports multiplexing several transactions over one connection.
```
it is not the concept that does so but the actual thing---a stream
